contents/dataframes_performance.md

```diff
@@ -4,8 +4,237 @@ So far, we didn't think about making our `DataFrames.jl` code **fast**.
 Like everything in Julia, `DataFrames.jl` can be really fast.
 In this section, we will give some performance tips and tricks.
 
-```{=comment}
-- Functions with a shebang versus without (e.g., filter! versus filter)
-- df[!, col] versus df[:, col]
-- other things? Jose probably knows some more
+### Allocations
+
+Like we explained in @sec:function_bang functions that end with a bang `!` are a common pattern to denote functions that modify one or more of their arguments.
+In the context of high performance Julia code, this *means* that **functions with `!` do not perform any sort of allocation**.
+They just change in place any object that we supply as arguments.
+
+Almost all the `DataFrames.jl` functions that we've seen so far have a \"`!` twin\".
+For example, `filter` has a non allocating `filter!`, `select` has `select!`, `subset` has `subset!`, and so on.
+Notice that these functions **do not** return a `DataFrame`, but instead they **alter** the `DataFrame` that they act upon.
+Additionally, there are some functions that do not have a `!` counterpart.
+For example, all the `join`s, by logical purposes, *cannot* be done without allocations.
+Thus, we don't have any `join!` functions in `DataFrames.jl`.
+
+If you want the highest speed and performance in your code you should definitely use the `!` functions instead of regular `DataFrames.jl` functions.
+
+Let's go back to the example of the `select` function in the beginning of @sec:select.
+Here is the responses `DataFrame`:
+
+```jl
+sco("responses()"; process=without_caption_label)
+```
+
+Now, let's perform the selection with the `select` function, like we did before:
+
+```jl
+s = """
+    # allocating function # hide
+    select(responses(), :id, :q1)
+    """
+sco(s, process=without_caption_label)
+```
+
+And here is the non allocating function:
+
+```jl
+s = """
+    # non allocarting function # hide
+    select!(responses(), :id, :q1)
+    """
+sco(s; process=without_caption_label)
+```
+
+Now let's see how they perform with the `@allocated` macro:
+
+```jl
+s = """
+    # allocation # hide
+    df = responses()
+    @allocated select(df, :id, :q1)
+    """
+sco(s; process=string, post=plainblock)
+```
+
+```jl
+s = """
+    # non allocation # hide
+    df = responses()
+    @allocated select!(df, :id, :q1)
+    """
+sco(s; process=string, post=plainblock)
+```
+
+As we can see `select!` makes less allocations that `select` so it should be faster while consuming less memory.
+
+### Copying vs Not Copying Columns
+
+There are **two ways to access a DataFrame column**.
+They differ in how they are accessed: one creates a "view" to the column without copying and the other creates a whole new column by copying the original column.
+
+The first is the regular dot `.` operator followed by the column name, like in `df.Col`.
```

Comment:
```suggestion
The first is the regular dot `.` operator followed by the column name, like in `df.col`.
```
Below, we used lowercase columns, so maybe here too?

Reply:
Actually we need to change it all then.